django_emarsys/util.py

```diff
@@ -0,0 +1,97 @@
+
+    if not getattr(settings, 'EMARSYS_BASE_URI', None):
+        log.warning("no EMARSYS_BASE_URI setting set")
+
+    events = getattr(settings, 'EMARSYS_EVENTS', None)
+    if not events:
+        log.warning("no EMARSYS_EVENTS setting set")
+        return
+
+    VALID_ARGUMENT = re.compile(r'^[a-z][a-z_0-9]*$')
+    for event, data in events.items():
+        if not event:
+            log.warning("invalid event name: '{event}'".format(event=event))
+
+        names = set()
+        args = set()
+        for data in data:
+            if 'name' not in data:
+                log.error("missing data key for event '{event}': 'name'"
+                          .format(event=event))
+
+            else:
+                name = data['name']
+                if name in names:
+                    log.warning(
+                        "reused data name for event '{event}': '{name}'"
+                        .format(name=name, event=event))
+
+                if not name or not name.strip():
+                    log.warning(
+                        "invalid data name for event '{event}': '{name}'"
+                        .format(name=name, event=event))
+
+                names.add(name)
+
+            if 'arg' not in data:
+                log.error("missing data key for event '{event}': 'arg'"
+                          .format(event=event))
+            else:
+                arg = data['arg']
+                if arg in args:
+                    log.error(
+                        "reused data argument for event '{event}': '{arg}'"
+                        .format(arg=arg, event=event))
+
+                if not arg or not VALID_ARGUMENT.match(arg):
+                    log.error(
+                        "invalid data argument for event '{event}': '{arg}'"
+                        .format(arg=arg, event=event))
+
+                args.add(arg)
+
+            if 'model' not in data:
+                log.error("missing data key for event '{event}': 'model'"
+                          .format(event=event))
+            else:
+                model = data['model']
+                chunks = model.split('.')
+                if len(chunks) < 2:
+                    log.error("model for event '{event}' expected to be "
+                              "'<app_label>.<model_name>', got: '{model}'"
+                              .format(event=event, model=model))
+                else:
+                    try:
+                        get_model(model)
+                    except LookupError as e:
+                        log.error("bad model '{model}' for event "
+                                  "'{event}': {error}"
+                                  .format(model=model, event=event, error=e))
+
+            for key in data:
+                if key not in ('name', 'arg', 'model'):
+                    log.warning(
+                        "unexpected data key for event '{event}': '{key}'"
+                        .format(event=event, key=key))
```

Comment:
I believe this code can be written simpler. For example, at the data level, by comparing `{'name', 'arg', 'model'}` with `set(data)` you can test for missing and additional keys. Then at the event level, you can look at all names `{data['name'] for `

I just saw `for data in data`. Confusing.

I'll pretend it says `for datom in data` henceforth.

At the event level, you can look at all names `[datom['name'] for datom in data if 'name' in datom]` and handle invalid and duplicate names. Similar for args.